docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,434 @@
+field. This is useful to allow temporary drift for debugging specific devices. As soon
+as this field is flipped back to `fleet`, the device controller will roll back the device
+configuration to the fleet mandated configuration and remove any modifications made to the
+device spec.
+
+Devices set to `spec.managementMode: manual`, but with a `metadata.owner` fleet , must signal
+a device status condition to indicate that the device is not in sync with the fleet because of
+this.
+
+```yaml
+  - lastTransitionTime: "2024-04-30T15:06:20Z"
+    message: This device is not in sync with the fleet
+    reason: Manually Managed
+    status: False
+    type: Management
+```
+
+When a fleet owned device is moved back to fleet management the status condition should be removed.
+
+##### `/device/{name}` with managementMode: manual
+
+Devices in manual management mode are not managed by the fleet controller, the device
+controller will not attempt to reconcile the device spec with the fleet spec.
+
+In this mode, the device.spec is fully writtable through this API.
+
+If the device is selected by a fleet, the `spec.managementMode` can be flipped to `fleet`, and the
+device controller will start managing the device again.
+
+While devices are managed by fleets, all floating tags are resolved to stable hashes, but
+when a device is managed manually, the floating tags can be used, see the next sub-section
+for additional details.
+
+###### Expected behaviour regarding external references with floating tags
+
+Changes on external references is monitored and acted on (git branches,
+device images, secrets), the device controller should re-evaluate the device
+spec and apply the changes to the rendered version which will be exposed
+on the `/device/{name}/rendered` endpoint.
+
+Several mechanisms can be used to track external references:
+* Polling
+* Webhooks
+
+#### `/device/{name}/rendered` (device side)
+This endpoint is used by the device agent to retrieve the rendered device spec, this endpoint is
+read-only, each device can only access it's own device rendered spec based on the device
+certificate identification, an administrator can access any device rendered spec for debug purposes.
+
+The agent can pass two parameters: `version`and `owner` if those haven't changed the API will return
+204 No Content, if the version or fleet has changed the API will return the rendered spec.
+
+Change proposed:
+* The parameter name `fleet` is changed to `owner` to be consistent with the metadata field.
+
+Behavior changes:
+* If a device configuration is not yet rendered for some reason, FC will queue the device for rendering and instruct
+  the device to try again later by returning 202 Accepted.
+
+The administrator can find the fully rendered device spec by using the following cmdline command:
+```shell
+flightctl get device/xxxxxxx --rendered
+```
+
+The JSON blob configuration exposed to the device in the /device/{name}/rendered endpoint
```

Comment:
We should switch back the returned format from RenderedDeviceSpec to DeviceSpec while we’re at it.

Reply:
We cannot use plain DeviceSpec since it contains many options the device cannot process (k8s secret, git secret, of future external refs).

Reply:
Can this not be a convention between service and agent to just use inlined config types on that interface (like with MCS/MCD)?

Reply:
If you are worried about the nested JSON format, we could expose a single ignition file in yaml.

BTW I think there is not a lot of benefit in discussing this part of the API. IMHO the important part is the frontend, and expected behavior, this could very well end up being another/a binary protocol

Reply:
You're right, this RFE should only focus on the user-facing API, and this is about the device-facing API that users shouldn't make assumptions on.

Just to clarify: The nested JSON isn't the concern; it was the same before in both FC and OCP. And by using the `-o json` flag in the CLI there also wasn't a concern filtering through `jq` if necessary. No, the point was to remain future-proof as the configuration format may change from Ignition to ConfigMaps and Secrets and potentially other items for app management, too, and to keep the same approach as MCO.